docs/riscv.rst

```diff
@@ -0,0 +1,257 @@
+   +    rgb_write((100000/64000)-1, LEDDBR);
+        int i = 0;
+        while (1) {
+            color_wheel(i++);
+
+What this does is increase the LED blink rate from 250 Hz to a much
+higher value. Compile this and load it again with
+``dfu-util -D riscv-blink.bin``. The blink rate should appear solid,
+because it’s blinking too quickly to see.
+
+Debugging RISC-V Code
+~~~~~~~~~~~~~~~~~~~~~
+
+Because we have ``peek`` and ``poke``, and because the USB bridge is a
+bus master, we can actually halt (and reset!) the CPU over the USB
+bridge. We can go even further and attach a full debugger to it!
+
+To start with, run ``wishbone-tool -s gdb``:
+
+.. prompt:: bash $ auto
+
+   $ wishbone-tool -s gdb
+   INFO [wishbone_tool::usb_bridge] opened USB device device 008 on bus 001
+   INFO [wishbone_tool::server] accepting connections on 0.0.0.0:1234
+
+In a second window, run gdb on ``riscv-blink.elf``:
+
+.. prompt:: bash $ auto
+
+   $ riscv64-unknown-elf-gdb riscv-blink.elf -ex 'target remote localhost:1234'
+   GNU gdb (GDB) 8.2.90.20190228-git
+   Copyright (C) 2019 Free Software Foundation, Inc.
+   License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
+   This is free software: you are free to change and redistribute it.
+   There is NO WARRANTY, to the extent permitted by law.
+   Type "show copying" and "show warranty" for details.
+   This GDB was configured as "--host=x86_64-w64-mingw32 --target=riscv64-unknown-elf".
+   Type "show configuration" for configuration details.
+   For bug reporting instructions, please see:
+   <http://www.gnu.org/software/gdb/bugs/>.
+   Find the GDB manual and other documentation resources online at:
+       <http://www.gnu.org/software/gdb/documentation/>.
+
+   For help, type "help".
+   Type "apropos word" to search for commands related to "word"...
+   Reading symbols from .\riscv-blink.elf...
+   Remote debugging using localhost:1234
+   csr_writel (addr=3758106660, value=1) at ./include/hw/common.h:41
+   41              *((volatile uint32_t *)addr) = value;
+   (gdb)
+
+If we run ``bt`` we can get a backtrace, and chances are that we landed
+in an ``msleep`` function:
+
+.. code:: gdb
+
+   (gdb) bt
+   #0  0x2004014c in csr_readl (addr=3758106664) at ./include/hw/common.h:46
+   #1  timer0_value_read () at ./include/generated/csr.h:242
+   #2  0x200401dc in msleep (ms=ms@entry=80) at ./include/hw/common.h:41
+   #3  0x20040074 in main () at ./src/main.c:45
+   (gdb)
+
+We can insert breakpoints, step, continue execution, and generally debug
+the entire system. We can even reset the program by running
+``mon reset``.
+
+Further RISC-V experiments
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+There is an additional RISC-V demo in the workshop. The
+``riscv-usb-cdcacm`` directory contains a simple USB serial device that
+simply echoes back any characters that you type, incremented by 1. This
+is a good way to get started with an interactive terminal program, or
+logging data via USB serial.
```

Comment:
Note that this is a hyper specialised program and only works on Foboot 1.x. For foboot-2.0, users can refer to the example code inside tinyusb.